django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,139 @@
+from django.db.models import Q
+from typing import Optional
+from main_app.models import (
+    Source,
+    Chant,
+    Sequence,
+)
+from users.models import User
+
+
+def user_can_edit_chants_in_source(user: User, source: Optional[Source]) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_can_proofread_chants_in_source(user: User, source: Source) -> bool:
+    """
+    Checks if the user can access the proofreading page of a given Source.
+    Used in ChantProofreadView.
+    """
+    source_id = source.id
+    if user.is_anonymous:
+        return False
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+
+    return (user_is_project_manager) or (user_is_editor and user_is_assigned_to_source)
+
+
+def user_can_view_unpublished_source(user: User, source: Source) -> bool:
+    """
+    Checks if the user can view an unpublished Source on the site.
+    Used in ChantDetail, SequenceDetail, and SourceDetail views.
+    """
+    user_is_authenticated: bool = user.is_authenticated
+    return (source.published) or (user_is_authenticated)
+
+
+def user_can_view_unpublished_chant(user: User, chant: Chant) -> bool:
+    """
+    Checks if the user can view a Chant belonging to an unpublished Source on the site.
+    Used in ChantDetail, SequenceDetail, and SourceDetail views.
+    """
+    source = chant.source
+    user_is_authenticated: bool = user.is_authenticated
+    return (source is not None) or (source.published) or (user_is_authenticated)
```

Comment:
But source will never be None - based on the docstring, this function accepts only chants that belong to unpublished sources. Checking for `source is not None` only makes sense for a function named `user_can_view_chant`

You get the point.